[ENH] Test & Score: Add comparison of models

janezd opened the pull request:
Solves #3891.

##### Includes
- [X] Code changes
- [x] Tests
- [x] Documentation

ajdapretnar commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -275,13 +282,108 @@ def __init__(self):
+        box.layout().addWidget(QLabel(
+            "<small>Table shows probabilities that the score for the model in "
+            "the row is higher than that of the model in the column. "
+            "Small numbers show the probability that the difference is"
```
space after final 'is'

ajdapretnar replied:
That said, that should probably be in the message above, not fixed below the table.

janezd replied:
I added the space, but would prefer having this text visible.

ajdapretnar commented:
This change makes Test&Score even more complicated. Is there a way to support this functionality outside of this widget? Perhaps as something that comes after Test&Score?

ajdapretnar commented:
I applaud the documentation commit. 👏 Very diligent.

janezd commented:
> I applaud the documentation commit.

I pushed two commits. One adds a long-sought Bayesian test for comparison of classifiers in some 100 lines of reasonably decent code (not the best I ever wrote, but it's OK) accompanied with 150 lines of tests that systematically check all branches. I worked on it for, say, 2 days. The second commit adds a few sentences I wrote in ten minutes while waiting for some check-up in a hospital. And you applaud the latter commit?! What an encouragement!

Anyway, I wanted to ask you for an updated screenshot with two new numbers -- assuming the layout is confirmed (but save the workflow, just for a case).

ajdapretnar commented:
I clicked through the widget and it works well. A couple of comments, though:

1. The width of the columns is dependent on the learner name, so short learns get squeezed into substandard cells, while long names get really comfy. 😁 I propose minimum column width + eliding longer learner names.
<img width="1233" alt="Screen Shot 2020-01-10 at 10 53 10" src="https://user-images.githubusercontent.com/12524972/72144690-3034c900-3399-11ea-928c-0eda64a994ab.png">

2. When a single learner is on the input, the bottom part should probably be disabled.
<img width="707" alt="Screen Shot 2020-01-10 at 10 54 23" src="https://user-images.githubusercontent.com/12524972/72144793-683c0c00-3399-11ea-8bc5-231f0ea52453.png">

3. Names of learners, both in rows and in columns, appear to be clickable, even though they're not. This can probably be disabled somewhere, so the GUI is not misleading.

4. [IDEA] How about coloring the cells according to the score? Say with light blue and the stronger the color the higher the score?

janezd commented:
I implemented 1-3, but I won't do the coloring. I agree this would be nice, but it requires too much machinery (going to models and defining delegates to either paint HTMLs or do the painting the hard way...) or it would be quick and dirty.

@VesnaT, you can review this as it is.

VesnaT commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -275,13 +283,53 @@ def __init__(self):
+        gui.checkBox(hbox, self, "use_rope",
+                     "Negligible difference: ",
+                     callback=self.update_comparison_table)
+        gui.lineEdit(hbox, self, "rope", validator=QDoubleValidator(),
```
Why is this not a spinbox? 
It should probably disabled when `use_rope` is not checked.

janezd replied:
It's not a spin box because it has no defined range. It can refer to AUC, that is, between 0 and 1, or it can be RMSE, which is between 0 and infinity -- it can easily be 100000.

Disabling it would make sense, I'll do that.

janezd replied:
I added disabling but didn't like it. Let's let the user change the line edit first and then enable it, if (s)he wishes.

I added a method `_on_use_rope_changed`. You can add a line `self.controls.rope.setEnabled(self.use_rope)` and see for yourself that you won't like it. :)

janezd replied:
This is a won't fix. :)

VesnaT commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -562,6 +620,91 @@ def update_stats_model(self):
+                    p0, rope, p1 = baycomp.two_on_single(
+                        row_scores, col_scores, self.rope)
+                    self._set_cell(table, row, col,
+                                   f"{p0:.3f}<br/><small>{rope:.3f})</small>",
```
`)` in 68. column should be removed.

janezd replied:
Nice. Fixed.

VesnaT commented on Orange/widgets/evaluate/tests/test_owtestlearners.py:
```diff
@@ -391,6 +399,204 @@ def test_no_pregressbar_warning(self):
+        baycomp.two_on_single.assert_called()
+        baycomp.two_on_single.reset_mock()
+
+    @patch("baycomp.two_on_single", Mock(wraps=baycomp.two_on_single))
```
No need to patch the `baycomp.two_on_single` for this test.

VesnaT commented on Orange/widgets/evaluate/tests/test_owtestlearners.py:
```diff
@@ -391,6 +399,204 @@ def test_no_pregressbar_warning(self):
+        self.get_output(self.widget.Outputs.evaluations_results, wait=5000)
+        self.assertTrue(w.comparison_table.isEnabled())
+
+    @patch("baycomp.two_on_single", Mock(wraps=baycomp.two_on_single))
```
Same as above.

janezd commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -562,6 +620,91 @@ def update_stats_model(self):
+        return slots
+
+    def _scores_by_folds(self, slots):
+        scorer = self.scorers[self.comparison_criterion]()
```
Can it?

VesnaT replied:
I just noticed `self.comparison_criterion` setting does not work.
But if it worked and if someone removes scores from scoring module, it probably can.

janezd replied:
It is now a schema-only setting, with pending. I added an out-of-range check. If scorers are added or removed, the setting can still be wrong (because it's an index) but it won't crash. I don't find this issue too important.

janezd commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -470,9 +521,16 @@ def shuffle_split_changed(self):
         self.__update()
 
+    def _update_comparison_enabled(self):
+        self.comparison_table.setEnabled(
```
I like disabling it because it shows the user that it's intentionally blank. Otherwise it looks like a bug when the upper table is filled and this one isn't (e.g. when using Leave one out). Hiding would also be an option, though I like disabling better -- like "something could be here, but currently isn't because I can't compute it in this situation".

I can disable it when there is no data. But in this case we should do the same with the above table, I suppose. We need to discuss this.

janezd replied:
Both views now disable under same conditions.

janezd commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -275,13 +283,53 @@ def __init__(self):
+        header.setSectionResizeMode(QHeaderView.ResizeToContents)
+        avg_width = self.fontMetrics().averageCharWidth()
+        header.setMinimumSectionSize(8 * avg_width)
+        header.setMaximumSectionSize(15 * avg_width)
```
That is, we can have fixed size. Alternatively, we can stretch all sections. What would you prefer? I prefer stretching. Can you please try `header.setSectionResizeMode(QHeaderView.Stretch)` (in line 318) and tell me whether you like it. It is a bit of a stretch for just two methods, while with three it's already OK.

VesnaT replied:
I like it for three method. I don't like it for two and a single method, and it makes it inconsistent with the evaluation table.

janezd replied:
Now it switches between Stretch and Fixed depending upon number of methods. Fixed has a default size of (about the same) as Logistic regression above (15 average chars). Stretch has a minimum size so that it starts scrolling when there are too many methods.

janezd commented on Orange/widgets/evaluate/owtestlearners.py:
```diff
@@ -611,6 +755,8 @@ def _invalidate(self, which=None):
                         item.setData(None, Qt.DisplayRole)
                         item.setData(None, Qt.ToolTipRole)
 
+        self.comparison_table.clearContents()
```
Fixed.

janezd replied:
Not it's really fixed. Probably.

VesnaT commented:
Why is the Evaluation Results table now disabled when one learner is on the input?

Right before Model comparison in performed, all the scores are shown in the left upper cell of the Model comparison table for a brief moment.

janezd commented:
> Why is the Evaluation Results table now disabled when one learner is on the input?

Because I'm stupid, but I've fixed that. (Now I'm smart.)

> Right before Model comparison in performed, all the scores are shown in the left upper cell of the Model comparison table for a brief moment.

Interesting. Fixed, too.
